processing/src/main/java/io/druid/query/search/SearchQueryRunner.java

```diff
@@ -330,4 +202,171 @@ public SearchHit apply(Map.Entry<SearchHit, MutableInt> input)
+        if (capabilities == null) {
+          continue;
+        }
+
+        if (capabilities.hasBitmapIndexes()) {
+          bitmapDims.add(spec);
+        } else {
+          nonBitmapDims.add(spec);
+        }
+      }
+    } else {
+      // no QueryableIndex available, so nothing has bitmaps
+      nonBitmapDims.addAll(dimsToSearch);
+    }
+  }
+
+  private void processNonBitmapDims(
+      SearchQuery query,
+      final StorageAdapter adapter,
+      Filter filter,
+      Interval interval,
+      final int limit,
+      final List<DimensionSpec> nonBitmapDims,
+      final SearchQuerySpec searchQuerySpec,
+      final Object2IntRBTreeMap<SearchHit> retVal
+  )
+  {
+    final Sequence<Cursor> cursors = adapter.makeCursors(filter, interval, query.getGranularity(), query.isDescending());
+
+    cursors.accumulate(
+        retVal,
+        new Accumulator<Object2IntRBTreeMap<SearchHit>, Cursor>()
+        {
+          @Override
+          public Object2IntRBTreeMap<SearchHit> accumulate(Object2IntRBTreeMap<SearchHit> set, Cursor cursor)
+          {
+            if (set.size() >= limit) {
+              return set;
+            }
+
+            List<QueryDimensionInfo> dimInfoList = Lists.newArrayList();
+            for (DimensionSpec dim : nonBitmapDims) {
+              dimInfoList.add(getDimInfoFromSpec(dim, adapter, cursor));
+            }
+
+            while (!cursor.isDone()) {
+              for (QueryDimensionInfo dimInfo : dimInfoList) {
+                dimInfo.queryHelper.updateSearchResultSet(
+                    dimInfo.outputName,
+                    dimInfo.selector,
+                    searchQuerySpec,
+                    limit,
+                    set
+                );
+
+                if (set.size() >= limit) {
+                  return set;
+                }
+              }
+
+              cursor.advance();
+            }
+
+            return set;
+          }
+        }
+    );
+  }
+
+  private void processBitmapDims(
+      QueryableIndex index,
+      Filter filter,
+      Interval interval,
+      List<DimensionSpec> bitmapDims,
+      SearchQuerySpec searchQuerySpec,
+      int limit,
+      final Object2IntRBTreeMap<SearchHit> retVal
+  )
+  {
+    final BitmapFactory bitmapFactory = index.getBitmapFactoryForDimensions();
+
+    final ImmutableBitmap baseFilter =
+        filter == null ? null : filter.getBitmapIndex(new ColumnSelectorBitmapIndexSelector(bitmapFactory, index));
+
+    ImmutableBitmap timeFilteredBitmap;
+    if (!interval.contains(segment.getDataInterval())) {
+      MutableBitmap timeBitmap = bitmapFactory.makeEmptyMutableBitmap();
+      final Column timeColumn = index.getColumn(Column.TIME_COLUMN_NAME);
+      try (final GenericColumn timeValues = timeColumn.getGenericColumn()) {
+
+        int startIndex = Math.max(0, getStartIndexOfTime(timeValues, interval.getStartMillis(), true));
+        int endIndex = Math.min(
```

Comment:
Are you referring to the endIndex calculation specifically, i.e., using `startIndex + timeValues.length() - 1` as the first argument to Math.min() instead of just `timeValues.length() - 1`?

If so, I think this is correct as-is.

`startIndex + timeValues.length() -1" could "overshoot" and be an invalid bitmap index position when startIndex > 0, although I don't think it would break anything either since `getStartIndexOfTime(timeValues, interval.getEndMillis(), false) - 1` would at most return `timeValues.length() -1` when interval.getEndMillis() is outside of the timestamp range in the timeColumn, or would return some index < timeValues.length -1 when interval.getEndMillis() is within the timeColumn's timestamp range

In any case, the index range calculation here is unchanged from when it was introduced in https://github.com/druid-io/druid/pull/2903, so I feel like tweaking this area which isn't directly related to the type-specific interface move is out of scope for this PR